models/User.php

```diff
@@ -145,22 +160,95 @@ public function validateAuthKey($authKey)
      */
     public function validatePassword($password)
     {
-        return $this->password === $password;
+        return Yii::$app->security->validatePassword($password, $this->password_hash);
+    }
+
+    public function login() {
+        $this->token = $this->generateToken();
+        return $this->save();
+    }
+
+    /**
+     * @return bool
+     * @throws \yii\base\Exception
+     */
+    public function register() {
+        $this->password_hash = Yii::$app->security->generatePasswordHash($this->password);
+        $this->token = $this->generateToken();
+        return $this->save();
+    }
+
+    /**
+     * @return string
+     * @throws InvalidConfigException
+     */
+    protected function generateToken() {
+        $signer = new Sha256();
+
+        if (!isset(\Yii::$app->jwt)) {
+            throw new InvalidConfigException('You must setup JWT component');
+        }
+
+        $token = \Yii::$app->jwt->getBuilder()
+            ->setExpiration(time() + self::TOKEN_EXPIRATION_TIMEOUT)
+            ->set('user_id', $this->id)
+            ->sign($signer, \Yii::$app->jwt->key)
+            ->getToken();
+        return strval($token);
     }
 
     /**
      * @return \yii\db\ActiveQuery
      */
     public function getFollows()
     {
-        return $this->hasMany(Follow::className(), ['follow_id' => 'id']);
+        return $this->hasMany(Follow::class, ['follow_id' => 'id']);
     }
 
     /**
      * @return \yii\db\ActiveQuery
      */
     public function getUsers()
     {
-        return $this->hasMany(Follow::className(), ['user_id' => 'id']);
+        return $this->hasMany(Follow::class, ['user_id' => 'id']);
+    }
+
+    /**
+     * @param User $user
+     */
+    public function follow($user) {
+        $isAlreadyFollowing = Follow::find()
+            ->andWhere(['follower_id' => $this->id, 'followed_id' => $user->id])
```

Comment:
her you can use `->where(['follower_id' => $this->id, 'followed_id' => $user->id])`

Reply:
Replace "where" with "andWhere" is good practice if Activerecord::find() already returns ActiveQuery with "where" command. But in this case it's not necessary, so I replace back with where()